ratbag_emu/device.py

```diff
@@ -0,0 +1,272 @@
+                'howling', 'squawking', 'cheering', 'warbling',
+                'thundering', 'booming', 'blustering', 'humming',
+                'crying', 'bawling', 'roaring', 'raging',
+                'chanting', 'crooning', 'murmuring', 'bellowing',
+                'wailing', 'weeping', 'screaming', 'yelling',
+                'yodeling', 'singing', 'honking', 'hooting',
+                'whispering', 'hollering',
+        ]
+
+        name = device_names[random.randint(0, len(device_names)-1)]
+        attr = device_attr[random.randint(0, len(device_attr)-1)]
+        return '-'.join([attr, name])
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def info(self):
+        return self._info
+
+    @property
+    def rdescs(self):
+        return self._rdescs
+
+    @property
+    def actuators(self):
+        return self._actuators
+
+    @actuators.setter
+    def actuators(self, val):
+        # Make sure we don't have actuators which will act on the same keys
+        seen = []
+        for keys in [a.keys for a in val]:
+            for el in keys:
+                assert el not in seen
+                seen.append(el)
+
+        self._actuators = val
+
+    def destroy(self):
+        for endpoint in self.endpoints:
+            endpoint.destroy()
+
+    def transform_action(self, data: Dict[str, Any]):
+        '''
+        Transforms high-level action according to the actuators
+
+        A high-level action will have the x, y values in mm. This values will
+        be converted to pixel by the device actuators (in this case, the
+        sensor/dpi actuator)
+
+        :param action:  high-level action
+        '''
+        hid_data: Dict[str, Any] = {}
+
+        for actuator in self.actuators:
+            hid_data.update(actuator.transform(data.copy()))
+
+        return hid_data
+
+    def send_hid_action(self, action: object):
+        '''
+        Sends a HID action
+
+        We assume there's only one endpoint for each type of action (mouse,
+        keyboard, button, etc.) so we send the action to all endpoints. The
+        endpoint will only act on the action if it supports it.
+
+        :param action:  HID action
+        '''
+        for endpoint in self.endpoints:
+            # FIXME: global data (0x11)?
+            endpoint.send(endpoint.create_report(action, 0x11))
+
+    def simulate_action(self, action: Dict[str, Any], type: int = None):
+        '''
+        Simulates action
+
+        Translates physical values according to the device properties and
+        converts action into HID reports.
+
+        :param action:  high-level action
+        :param type:    HID report type
+        '''
+
+        packets: Dict[int, EventData] = {}
+
+        report_count = int(round(ms2s(action['duration']) * self.report_rate))
+
+        if not report_count:
+            report_count = 1
+
+        # XY movement
+        if action['type'] == 'xy':
```

Comment:
we should probably use enums for the action types, so we don't mess up with the strings.